### PR TITLE
vim: Add a setting to change the Vim mode indicator position on the status bar from right to left

### PR DESCRIPTION
**Pull request by helgemahrt**
Add the first request on issue #14093

Release Notes:

- Added a setting to change the Vim mode indicator position on the status bar from right to left

### CODE REVIEW

File: crates/project/src/lsp_store.rs [hunk 1]
```diff
@@ -4906,96 +4907,97 @@ impl LspStore {
         let buffer = buffer.read(cx);
         let file = File::from_dyn(buffer.file())?;
         let abs_path = file.as_local()?.abs_path(cx);
-        let uri = lsp::Url::from_file_path(abs_path).unwrap();
-        let next_snapshot = buffer.text_snapshot();
+        if let Ok(uri) = lsp::Url::from_file_path(abs_path) {
```
Comment (maxdeviant) [inline crates/project/src/lsp_store.rs:4910]: How are these changes in the `LspStore` related to the Vim mode indicator?

Comment (helgemahrt) [reply crates/project/src/lsp_store.rs:4910]: Not directly, actually. This fixed failing unit tests on Windows. Should I pull this out into a separate PR?

Comment (maxdeviant) [reply crates/project/src/lsp_store.rs:4910]: > Should I pull this out into a separate PR?

That would be great, thank you!

Comment (helgemahrt) [reply crates/project/src/lsp_store.rs]: Done :)

**Comment by helgemahrt** [comment]
@maxdeviant better now?

**Comment by ConradIrwin** [comment]
@helgemahrt Thanks for this!

We should put the mode indicator to the left of the diagnostics, otherwise when the diagnostics update the mode indicator jumps  around. The tricky bit is that the diagnostics indicator currently provides the vertical line, so we'll need to to fix how that is rendered.

The only reason this is on the right today is this issue, so I'd also be in favor of just moving this to the left unconditionally and saving ourselves a probably unnecessary setting.

**Comment by helgemahrt** [comment]
IMO it's probably a good idea to have it configurable - there may be people who prefer it on the right side. However, maybe the better approach would be to make the order of all elements of the status bar configurable?

For now, I added a change that ensures that the diagnostics indicator stays the right-most element on the left side of the status bar.

**Comment by helgemahrt** [comment]
Hang on, I made a mess in this branch when I removed the other commits. I'll create a clean branch/PR and link it here.

**Comment by ConradIrwin** [comment]
@helgemahrt I think it needs to be the *left-most* position on the left of the status bar. I don't want it to move when the diagnostics updates.

**Comment by helgemahrt** [comment]
I think I managed to recover this branch. 😅

> @helgemahrt I think it needs to be the _left-most_ position on the left of the status bar. I don't want it to move when the diagnostics updates.

Currently, the diagnostics indicator is the right-most item on the left side, with the vertical separator to the left:
image

My latest change ensures it stays that way, but, TBH, it feels a bit hacky.
image

I just realized I missed ensuring that the Vim mode indicator stays behind the diagnostics indicator on startup as well. Will fix that in a moment.

**Comment by helgemahrt** [comment]
I just realized that there's also the activity indicator. Now I'm thinking that maybe it'd be better to scrap this PR and rethink whether and how the status bar should be configurable. And maybe put the vim mode indicator to the left side for the time being, like you suggested. WDYT?

**Comment by ConradIrwin** [comment]
Sounds good! If you're up for it I think moving to the left of all the things would be a positive development

**Comment by helgemahrt** [comment]
Created #22269 to move the indicator. Closing this PR.

**Pull request closed without merging**